File: bowershub-ai/backend/models/conversation.py
"""Pydantic models for conversations and messages."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from pydantic import BaseModel, Field


class ConversationCreate(BaseModel):
    workspace_id: int
    title: Optional[str] = None


class ConversationUpdate(BaseModel):
    title: Optional[str] = Field(None, max_length=200)
    is_archived: Optional[bool] = None


class ConversationListItem(BaseModel):
    id: int
    workspace_id: int
    title: Optional[str] = None
    parent_id: Optional[int] = None
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    message_count: int = 0


class ConversationResponse(BaseModel):
    id: int
    workspace_id: int
    user_id: int
    title: Optional[str] = None
    parent_id: Optional[int] = None
    branch_point_msg: Optional[int] = None
    is_archived: bool
    created_at: datetime
    updated_at: datetime
    messages: List["MessageResponse"] = []


class MessageResponse(BaseModel):
    id: int
    conversation_id: int
    role: str
    content: str
    attachments: List[Dict[str, Any]] = []
    model_used: Optional[str] = None
    routing_layer: Optional[str] = None
    input_tokens: Optional[int] = None
    output_tokens: Optional[int] = None
    cost_usd: Optional[float] = None
    metadata: Dict[str, Any] = {}
    created_at: datetime


class SendMessageRequest(BaseModel):
    content: str = Field(..., min_length=1, max_length=10000)
    model: str = "auto"
    attachments: List[Dict[str, Any]] = []

File: bowershub-ai/frontend/src/schemas/conversation.ts
/**
 * Zod schemas for conversation/message wire types.
 *
 * These schemas are the single source of truth: the TS types are derived via
 * `z.infer`, so the compile-time shape and the runtime validation can never
 * drift apart. `.passthrough()` keeps unknown fields (a new backend column
 * won't be silently dropped). Consumers import the types from
 * `stores/conversation`, which re-exports them.
 */

import { z } from 'zod'

export const MessageSchema = z
  .object({
    id: z.number(),
    conversation_id: z.number(),
    role: z.enum(['user', 'assistant', 'system', 'tool_call', 'tool_result']),
    content: z.string(),
    attachments: z.array(z.any()),
    model_used: z.string().nullable(),
    routing_layer: z.string().nullable(),
    input_tokens: z.number().nullable(),
    output_tokens: z.number().nullable(),
    cost_usd: z.number().nullable(),
    metadata: z.record(z.string(), z.any()),
    created_at: z.string(),
  })
  .passthrough()

export const ConversationSchema = z
  .object({
    id: z.number(),
    workspace_id: z.number(),
    title: z.string().nullable(),
    parent_id: z.number().nullable(),
    is_archived: z.boolean(),
    created_at: z.string(),
    updated_at: z.string(),
    message_count: z.number(),
  })
  .passthrough()

export type Message = z.infer<typeof MessageSchema>
export type Conversation = z.infer<typeof ConversationSchema>
